cpp/src/arrow/array/util.cc

```diff
@@ -74,6 +74,177 @@ class ArrayDataWrapper {
   std::shared_ptr<Array>* out_;
 };
 
+class ArrayDataEndianSwapper {
+ public:
+  ArrayDataEndianSwapper(std::shared_ptr<ArrayData>& data, int64_t length)
+      : data_(data), length_(length) {}
+
+  Status SwapType(const DataType& type) { return VisitTypeInline(type, this); }
+
+  Status SwapChildren(std::vector<std::shared_ptr<Field>> child_fields) {
+    int i = 0;
+    for (const auto& child_field : child_fields) {
+      auto orig_data = data_;
+      auto orig_length = length_;
+      data_ = data_->child_data[i++];
+      length_ = data_->length;
+      RETURN_NOT_OK(SwapType(*child_field.get()->type()));
+      length_ = orig_length;
+      data_ = orig_data;
+    }
+    return Status::OK();
+  }
+
+  template <typename VALUE_TYPE>
+  Status SwapOffset(int index) {
+    if (data_->buffers[index] == nullptr) {
+      return Status::OK();
+    }
+    auto buffer = const_cast<VALUE_TYPE*>(
```

Comment:
This is definitely not safe at all. Another approach should be found IMHO. Worst case we reallocate, which I agree isn't desirable...

Reply:
I agree with safety. Since some of Arrow source files use `const_cast`, I also used it.   
This is the tradeoff between effectiveness and safety.

Reply:
As far as I can tell, we only use `const_cast` when invoking third-party (usually C) libraries that didn't bother being `const`-friendly.

Reply:
In this case anyway, we don't know what kind of buffer we are dealing with (it could be a private memory buffer, it could be a memory-mapped file, it could be shared with another library...).

Reply:
`gpu/cuda_memory.cc`, `io/file.cc`, and `io/memory.cc` are what I saw.   
If we should avoid the usage of `const_cast`, I agree. Should we reallocate the buffer?

Reply:
In the general case, we should certainly reallocate the buffer. I'm not sure if there are cases where we can avoid it, but our current APIs don't make it easy (a Buffer is a bit too abstracted for that).

In `gpu/cuda_memory.cc`, `io/file.cc` and `io/memory.cc`, the pointer is not dereferenced (the memory is not written to). Here, we are writing into the memory, so can't ignore `const`.

Reply:
I will address this issue tomorrow.

Reply:
@pitrou Since I was swamped with something in my company, I was late to address this. Now, I updated the code to reallocate the buffer. Could you please look at this again?

Reply:
@pitrou How about this change?

Reply:
Well, I still see a `const_cast` in `SwapOffset`. Otherwise the approach looks fine :-)

Reply:
Sorry for my bad. I updated `SwapOffset`.  The current failures in R are not related to this update.